src/vcpkg/commands.ci.cpp

```diff
@@ -357,57 +412,40 @@ namespace vcpkg
         auto& var_provider = *var_provider_storage;
 
         const ElapsedTimer timer;
-        // Install the default features for every package
-        std::vector<FullPackageSpec> all_default_full_specs;
-        for (auto scfl : provider.load_all_control_files())
-        {
-            all_default_full_specs.emplace_back(
-                PackageSpec{scfl->to_name(), target_triplet},
-                InternalFeatureSet{FeatureNameCore.to_string(), FeatureNameDefault.to_string()});
-        }
-
-        struct RandomizerInstance : GraphRandomizer
-        {
-            virtual int random(int i) override
-            {
-                if (i <= 1) return 0;
-                std::uniform_int_distribution<int> d(0, i - 1);
-                return d(e);
-            }
+        auto ci_specs = calculate_ci_specs(exclusions_map, target_triplet, provider);
 
-            std::random_device e;
-        } randomizer_instance;
-        GraphRandomizer* randomizer = nullptr;
+        Optional<CiRandomizer> randomizer;
         if (Util::Sets::contains(options.switches, SwitchXRandomize))
         {
-            randomizer = &randomizer_instance;
+            randomizer.emplace();
         }
 
         PackagesDirAssigner packages_dir_assigner{paths.packages()};
         CreateInstallPlanOptions create_install_plan_options(
-            randomizer, host_triplet, UnsupportedPortAction::Warn, UseHeadVersion::No, Editable::No);
+            randomizer.get(), host_triplet, UnsupportedPortAction::Warn, UseHeadVersion::No, Editable::No);
         auto action_plan = compute_full_plan(
```

Comment:
It seems like there's still an unfortunate result when a user makes a small deviation outside a correct plan. For example, adding a dependency to an excluded package. While we know by construction that the resulting CI plan is "invalid", there's a bad constraint amplification problem where one bad constraint (use the excluded package) gets amplified into all of the excluded package's transitive constraints (that got it excluded in the first place).

Probably not worth solving right now, because it would require the dependency calculator to become aware of exclusions and ignore transitive constraints.